embassy-stm32/src/flash/mod.rs

```diff
@@ -0,0 +1,401 @@
+            pac::FLASH.cr().write(|w| w.set_pg(false));
+        }
+
+        ret
+    }
+
+    pub fn blocking_erase(&mut self, from: u32, to: u32) -> Result<(), Error> {
+        if to < from || to as usize > FLASH_END {
+            return Err(Error::Size);
+        }
+        if from as usize % PAGE_SIZE != 0 || to as usize % PAGE_SIZE != 0 {
+            return Err(Error::Unaligned);
+        }
+
+        self.clear_all_err();
+
+        for page in (from..to).step_by(PAGE_SIZE) {
+            #[cfg(any(flash_l0, flash_l1))]
+            unsafe {
+                pac::FLASH.pecr().modify(|w| {
+                    w.set_erase(true);
+                    w.set_prog(true);
+                });
+
+                write_volatile(page as *mut u32, 0xFFFFFFFF);
+            }
+
+            #[cfg(any(flash_wl, flash_wb, flash_l4))]
+            unsafe {
+                let idx = page / PAGE_SIZE as u32;
+
+                pac::FLASH.cr().modify(|w| {
+                    w.set_per(true);
+                    w.set_pnb(idx as u8);
+                    #[cfg(any(flash_wl, flash_wb))]
+                    w.set_strt(true);
+                    #[cfg(any(flash_l4))]
+                    w.set_start(true);
+                });
+            }
+
+            let ret: Result<(), Error> = self.blocking_wait_ready();
+
+            #[cfg(any(flash_wl, flash_wb, flash_l4))]
+            unsafe {
+                pac::FLASH.cr().modify(|w| w.set_per(false));
+            }
+
+            #[cfg(any(flash_l0, flash_l1))]
+            unsafe {
+                pac::FLASH.pecr().modify(|w| {
+                    w.set_erase(false);
+                    w.set_prog(false);
+                });
+            }
+
+            self.clear_all_err();
+            if ret.is_err() {
+                return ret;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn blocking_wait_ready(&self) -> Result<(), Error> {
+        loop {
+            let sr = unsafe { pac::FLASH.sr().read() };
+
+            if !sr.bsy() {
+                #[cfg(any(flash_wl, flash_wb, flash_l4))]
+                if sr.progerr() {
+                    return Err(Error::Prog);
+                }
+
+                if sr.wrperr() {
+                    return Err(Error::Protected);
+                }
+
+                if sr.pgaerr() {
+                    return Err(Error::Unaligned);
+                }
+
+                if sr.sizerr() {
+                    return Err(Error::Size);
+                }
+
+                #[cfg(any(flash_wl, flash_wb, flash_l4))]
+                if sr.miserr() {
+                    return Err(Error::Miss);
+                }
+
+                #[cfg(any(flash_wl, flash_wb, flash_l4))]
+                if sr.pgserr() {
+                    return Err(Error::Seq);
+                }
+                return Ok(());
+            }
+        }
+    }
+
+    fn clear_all_err(&mut self) {
+        unsafe {
+            pac::FLASH.sr().modify(|w| {
```

Comment:
maybe this can be done by simply writing 0?

Reply:
Will test it